torch/nn/parallel/distributed.py

```diff
@@ -136,13 +135,14 @@ def __init__(self, module, device_ids=None, output_device=None, dim=0):
             bucket_bytes_cap = 1 * MB
         bucket_bytes = bucket_bytes_cap  # to init the first bucket immediately
         for param_tuple in zip(*map(lambda m: m.parameters(), self._module_copies)):
-            if bucket_bytes >= bucket_bytes_cap:
-                self.bucket_sizes.append(0)
-                bucket_bytes = 0
-            self.bucket_sizes[-1] += 1
-            for p in param_tuple:
-                self.bucket_map[p] = len(self.bucket_sizes) - 1
-            bucket_bytes += p.numel() * p.element_size()
+            if param_tuple[0].requires_grad:
```

Comment:
The code could still be simplified and the whole diff could be two lines (I'd rather not move the logic around like you did below):
```python
if not param_tuple[0].requires_grad:
    continue
```
We should also give it some more thought. We will never check these flags again right now, and we should either make sure that they are never changed, or we will be missing gradient reductions for a subset of parameters which is really bad.

Reply:
What is happening now is also really bad (as detailed in #2533), and it is fundamental deficiency of how distributed data parallel is written now. Now, if you have some parameters that don't require grad, you still include them in the buckets, so all the buckets won't be filled per iteration, and reductions are happening at unpredictable moments in time.

Reply:
It is bad, but it will raise a proper error for sure. This solution can cause some of the parameters to be silently ignored. I agree the previous solution is a bit of a hack, but I don't think that piling more hacks on top of it will help us in the end. We should figure out a way to actually support params with different flags properly.

Reply:
No, it does not raise proper error! Ever since #2464, if a model has some parameters that don't require grads, the reductions would silently not be triggered at the right moments (and will be triggered at the wrong ones). Before #2464, the model that had parameters that did not require grads errored out.

Reply:
I'm all for figuring out proper solutions, but currently parameters that don't require grad are not supported in distributed, and release notes should not say otherwise https://github.com/pytorch/pytorch/releases (distributed and multi-gpu section)

Reply:
fixed the release notes.